blockindex/bloomfilterindexer.go

```diff
@@ -0,0 +1,253 @@
+			return err
+		}
+	default:
+		return err
+	}
+	return nil
+}
+
+// Stop stops the bloomfilter indexer
+func (bfx *bloomfilterIndexer) Stop(ctx context.Context) error {
+	return bfx.kvStore.Stop(ctx)
+}
+
+// Height returns the tipHeight from underlying DB
+func (bfx *bloomfilterIndexer) Height() (uint64, error) {
+	h, err := bfx.kvStore.Get(BlockBloomFilterNamespace, []byte(CurrentHeightKey))
+	if err != nil {
+		return 0, err
+	}
+	return byteutil.BytesToUint64(h), nil
+}
+
+// PutBlock processes new block by adding logs into rangebloomfilter, and if necessary, updating underlying DB
+func (bfx *bloomfilterIndexer) PutBlock(ctx context.Context, blk *block.Block) (err error) {
+	bfx.mutex.Lock()
+	defer bfx.mutex.Unlock()
+	bfx.addLogsToRangeBloomFilter(ctx, blk.Height(), blk.Receipts)
+	// commit into DB and update tipHeight
+	if err := bfx.commit(blk.Height(), bfx.calculateBlockBloomFilter(ctx, blk.Receipts)); err != nil {
+		return err
+	}
+	if blk.Height()%bfx.rangeSize == 0 {
+		bfx.curRangeBloomfilter, err = bloom.NewBloomFilter(2048, 3)
+		if err != nil {
+			return errors.Wrapf(err, "Can not create new bloomfilter")
+		}
+	}
+	return nil
+}
+
+// DeleteTipBlock deletes tip height from underlying DB if necessary
+func (bfx *bloomfilterIndexer) DeleteTipBlock(blk *block.Block) (err error) {
+	bfx.mutex.Lock()
+	defer bfx.mutex.Unlock()
+	height := blk.Height()
+	if err := bfx.delete(height); err != nil {
+		return err
+	}
+	if (height-1)%bfx.rangeSize == 0 {
+		bfx.curRangeBloomfilter = nil
+	} else {
+		bfx.curRangeBloomfilter, err = bfx.rangeBloomFilter(height - 1)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RangeBloomFilterSize returns the number of blocks that each rangeBloomfilter includes
+func (bfx *bloomfilterIndexer) RangeBloomFilterSize() uint64 {
+	bfx.mutex.RLock()
+	defer bfx.mutex.RUnlock()
+	return bfx.rangeSize
+}
+
+// BloomFilterByHeight returns the block-level bloomfilter which includes not only topic but also address of logs info by given block height
+func (bfx *bloomfilterIndexer) BloomFilterByHeight(height uint64) (bloom.BloomFilter, error) {
+	return bfx.blockBloomFilter(height)
+}
+
+// FilterBlocksInRange returns the block numbers by given logFilter in range (start, end]
+func (bfx *bloomfilterIndexer) FilterBlocksInRange(l *filter.LogFilter, start, end uint64) ([]uint64, error) {
+	bfx.mutex.RLock()
+	defer bfx.mutex.RUnlock()
+	blockNumbers := make([]uint64, 0)
+	if start%bfx.rangeSize != 0 {
+		return nil, errors.New("Invalid arguments")
+	}
+	queryHeight := start + bfx.rangeSize
+	for queryHeight <= end {
+		bigBloom, err := bfx.rangeBloomFilter(queryHeight)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get rangeBloomFilter from indexer by given height %d", queryHeight)
+		}
+		if l.ExistInBloomFilterv2(bigBloom) {
+			blocks := l.SelectBlocksFromRangeBloomFilter(bigBloom, queryHeight-bfx.rangeSize+1, queryHeight)
```

Comment:
if a block number is smaller than `start`, discard it. same for case larger than `end`

Reply:
I discarded if it is smaller than start, but for ```end```, I think I still have to handle this.